core/src/main/scala/org/bitcoins/core/gcs/GCS.scala

```diff
@@ -0,0 +1,227 @@
+    }
+
+    hashedItemsBuilder.result()
+  }
+
+  /**
+    * Converts num to unary (6 = 1111110)
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#golomb-rice-coding]]
+    *
+    * TODO: protect against large inputs which cause OutOfMemoryErrors and even larger ones which fail on toInt
+    */
+  def toUnary(num: UInt64): BitVector = {
+    if (num == UInt64.zero) {
+      bin"0"
+    } else {
+      /*
+       * We use the fact that 2^n - 1 = 111...1 (in binary) where there are n 1 digits
+       */
+      val binUnary = (BigInt(1) << num.toInt) - 1
+      val leftPadded = BitVector(binUnary.toByteArray)
+      val noPadding = dropLeftPadding(leftPadded)
+
+      noPadding.:+(false)
+    }
+  }
+
+  /**
+    * Encodes a hash into a unary prefix and binary suffix
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#golomb-rice-coding]]
+    */
+  def golombEncode(item: UInt64, p: UInt8): BitVector = {
+    val q = item >> p.toInt
+
+    val prefix = toUnary(q)
+
+    val pBits = item.bytes.toBitVector.takeRight(p.toInt)
+
+    prefix ++ pBits
+  }
+
+  /**
+    * Decodes an item off of the front of a BitVector by reversing [[GCS.golombEncode]]
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#golomb-rice-coding]]
+    */
+  def golombDecode(codedItem: BitVector, p: UInt8): UInt64 = {
+    @tailrec
+    def split(vec: BitVector, accum: UInt64): (UInt64, BitVector) = {
+      if (vec.head) {
+        split(vec.tail, accum + UInt64.one)
+      } else {
+        (accum, vec.tail)
+      }
+    }
+
+    val (q, pBits) = split(codedItem, UInt64.zero)
+
+    val sizeWithPadding = (8 - (p.toInt % 8)) + p.toInt
+
+    val pBitsAsBytes = {
+      val withoutRightPaddingOrData = pBits.take(p.toInt)
+      val withLeftPadding = withoutRightPaddingOrData.padLeft(sizeWithPadding)
+      withLeftPadding.toByteVector
+    }
+
+    (q << p.toInt) + UInt64.fromBytes(pBitsAsBytes)
+  }
+
+  @tailrec
+  private def dropLeftPadding(padded: BitVector): BitVector = {
+    if (padded.isEmpty || padded.head) {
+      padded
+    } else {
+      dropLeftPadding(padded.tail)
+    }
+  }
+
+  /** Returns the first hash gcs-encoded at the front of a BitVector, as well as the remaining BitVector */
+  private def golombDecodeItemFromSet(
+      encodedData: BitVector,
+      p: UInt8): (UInt64, BitVector) = {
+    val head = golombDecode(encodedData, p)
+
+    val prefixSize = (head >> p.toInt).toInt + 1
+
+    (head, encodedData.drop(prefixSize + p.toInt))
+  }
+
+  /**
+    * Decodes all hashes from golomb-encoded data, reversing [[GCS.encodeSortedSet]]
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#set-queryingdecompression]]
+    */
+  def golombDecodeSet(encodedData: BitVector, p: UInt8): Vector[UInt64] = {
+    @tailrec
+    def loop(
+        encoded: BitVector,
+        decoded: Vector[UInt64],
+        lastHash: UInt64 = UInt64.zero): Vector[UInt64] = {
```

Comment:
This default param seems like an easy foot gun, it doesn't seem to be that useful?